Proper r/w operations during a schema change with a column add/delete

vivekmenezes opened the pull request:
https://github.com/cockroachdb/cockroach/issues/2845

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3023)

<!-- Reviewable:end -->

vivekmenezes commented:
please review when you get a chance.

tamird commented:
Reviewed 10 of 12 files at r1.
Review status: 10 of 12 files reviewed at latest revision, 11 unresolved discussions.

---

<sup>**[sql/descriptor_mutation_test.go, line 21 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2SLnrUC8S0L--Qt3UL)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4f6e4d687f90bc4964f42e8069ffdc63fb0d0013/sql/descriptor_mutation_test.go#L21)):</sup>
We only import both packages in one other place, but we alias our sql package to "csql" there. Would you mind being consistent with that?
<!-- Reviewable comment -K2SLnrUC8S0L--Qt3UM:-1648812556 -->

---

<sup>**[sql/descriptor_mutation_test.go, line 28 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2SLwf5pPYWrVcS0Xh0)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4f6e4d687f90bc4964f42e8069ffdc63fb0d0013/sql/descriptor_mutation_test.go#L28)):</sup>
why this empty line?
<!-- Reviewable comment -K2SLwf5pPYWrVcS0Xh1:1745835411 -->

---

<sup>**[sql/descriptor_mutation_test.go, line 98 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2SPKYgrIaY40yK9izR)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4f6e4d687f90bc4964f42e8069ffdc63fb0d0013/sql/descriptor_mutation_test.go#L98)):</sup>
there needs to be a defer that undoes this.
<!-- Reviewable comment -K2SPKYgrIaY40yK9izS:-424093774 -->

---

<sup>**[sql/insert.go, line 141 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2SNk1IrByySN2J4t5l)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4f6e4d687f90bc4964f42e8069ffdc63fb0d0013/sql/insert.go#L141)):</sup>
I believe this is not the right level to be doing this at; we should really introduce accessor methods on the descriptors to do this correctly. At this point, you're going to allow a user to INSERT INTO foo VALUES (1,2,3) where the table has only two columns where the third is a DELETE_ONLY column - we should really emit an error instead.
<!-- Reviewable comment -K2SNk1IrByySN2J4t5m:931526383 -->

---

<sup>**[sql/parser/select.go, line 63 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2SOIgwjD9RoIe0LCSa)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4f6e4d687f90bc4964f42e8069ffdc63fb0d0013/sql/parser/select.go#L63)):</sup>
See my other comment about visibility accessors on the table descriptor. I believe those accessors would remove the need for this plumbing.
<!-- Reviewable comment -K2SOIgwjD9RoIe0LCSb:689576330 -->

---

<sup>**[sql/scan.go, line 298 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2SOZmbti5JOBkHU6VZ)** ([raw file](https://github.com/cockroachdb/cockroach/blob/e675a722519a8aa7fbf00d76edfecf5563f58281/sql/scan.go#L298)):</sup>
See my other comment about visibility accessors on the table descriptor.
<!-- Reviewable comment -K2SOZmbti5JOBkHU6V_:1382840569 -->

---

<sup>**[sql/structured.go, line 292 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2SP0NHvntyVZQqIXZ2)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4f6e4d687f90bc4964f42e8069ffdc63fb0d0013/sql/structured.go#L292)):</sup>
nit: slightly less susceptible to rot is: `util.Errorf("mutation in %s state: col %s, col-id %d", column.Mutation.State, column.Name, column.ID)`
<!-- Reviewable comment -K2SP0NHvntyVZQqIXZ3:1002467207 -->

---

<sup>**[sql/structured.proto, line 37 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2SMe0unsGGkg8hhL38)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4f6e4d687f90bc4964f42e8069ffdc63fb0d0013/sql/structured.proto#L37)):</sup>
make this comment a bit more explicit about how reads are restricted (i.e. by the very presence of a DescriptorMutation). Technically, `State` only restricts writes.
<!-- Reviewable comment -K2SMe0v1-KK05n3iWlq:182986354 -->

---

<sup>**[sql/structured.proto, line 40 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2SMVpuxABXj42ZXWLA)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4f6e4d687f90bc4964f42e8069ffdc63fb0d0013/sql/structured.proto#L40)):</sup>
nit: I think we use UNKNOWN elsewhere.
<!-- Reviewable comment -K2SMVpuxABXj42ZXWLB:1924128194 -->

---

<sup>**[sql/structured.proto, line 86 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2SN25pwaQb1lyM71Y5)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4f6e4d687f90bc4964f42e8069ffdc63fb0d0013/sql/structured.proto#L86)):</sup>
s/nil/null/ since this is proto style, not go style. also this comment should indicate that nil means the descriptor may be used for reads.
<!-- Reviewable comment -K2SN25pwaQb1lyM71Y6:-2016907351 -->

---

<sup>**[sql/update.go, line 134 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2SPduiayrVYbHqYjtb)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4f6e4d687f90bc4964f42e8069ffdc63fb0d0013/sql/update.go#L134)):</sup>
I don't think this is right; update shouldn't be able to see DELETE_ONLY columns.
<!-- Reviewable comment -K2SPduiayrVYbHqYjtc:-349743714 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 10 of 12 files reviewed at latest revision, 8 unresolved discussions.

---

<sup>**[sql/descriptor_mutation_test.go, line 21 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2SLnrUC8S0L--Qt3UL)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4f6e4d687f90bc4964f42e8069ffdc63fb0d0013/sql/descriptor_mutation_test.go#L21)):</sup>
done
<!-- Reviewable comment -K2SUycUqP3uebi0IbLX:677588487 -->

---

<sup>**[sql/descriptor_mutation_test.go, line 28 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2SLwf5pPYWrVcS0Xh0)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4f6e4d687f90bc4964f42e8069ffdc63fb0d0013/sql/descriptor_mutation_test.go#L28)):</sup>
fixed
<!-- Reviewable comment -K2SV0YsOjwmiuEbsltw:-467915377 -->

---

<sup>**[sql/descriptor_mutation_test.go, line 98 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2SPKYgrIaY40yK9izR)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4f6e4d687f90bc4964f42e8069ffdc63fb0d0013/sql/descriptor_mutation_test.go#L98)):</sup>
done
<!-- Reviewable comment -K2SVF1cXOVyOU7gDBoV:677588487 -->

---

<sup>**[sql/insert.go, line 141 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2SNk1IrByySN2J4t5l)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4f6e4d687f90bc4964f42e8069ffdc63fb0d0013/sql/insert.go#L141)):</sup>
That's certainly an approach. If we go down that route, we're completely turning off external reads and writes on a column with a mutation in any state. Which means that the states are not needed for column descriptors and are only needed by index descriptors. 

Use case:
Add a column: Add a column, wait until the column is completely inserted (default values are populated for instance), and the mutation is declared complete. Then allow operations that include the column.
Delete a column: Turn off use of the column by application code everywhere. Delete the column. During deletion, requests with the column are rejected.

What about a column with a constraint? Since we are turning off external reads/writes on a column during a schema change, the index automatically will not be populated from the outside world. 
<!-- Reviewable comment -K2SfLWZboIvI_avblGM:1092582800 -->

---

<sup>**[sql/structured.proto, line 37 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2SMe0unsGGkg8hhL38)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4f6e4d687f90bc4964f42e8069ffdc63fb0d0013/sql/structured.proto#L37)):</sup>
done
<!-- Reviewable comment -K2SWKxlB-_W0Mkd4CLu:677588487 -->

---

<sup>**[sql/structured.proto, line 86 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2SN25pwaQb1lyM71Y5)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4f6e4d687f90bc4964f42e8069ffdc63fb0d0013/sql/structured.proto#L86)):</sup>
fixed
<!-- Reviewable comment -K2SX9Yi9ZUnvGyGSWQf:-467915377 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 10 of 12 files reviewed at latest revision, 8 unresolved discussions.

---

<sup>**[sql/insert.go, line 141 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2SNk1IrByySN2J4t5l)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4f6e4d687f90bc4964f42e8069ffdc63fb0d0013/sql/insert.go#L141)):</sup>
I don't understand this assertion. Adding a column is exactly the same as adding an index, except for the special case of a NULL default, where the backfill is a no-op.
<!-- Reviewable comment -K2SpT5szDAalnK6vzNV:392323927 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 10 of 12 files reviewed at latest revision, 8 unresolved discussions.

---

<sup>**[sql/insert.go, line 141 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2SNk1IrByySN2J4t5l)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4f6e4d687f90bc4964f42e8069ffdc63fb0d0013/sql/insert.go#L141)):</sup>
I thought we could simplify column schema changes to the following state machines:

non-existent -> column-incomplete -> public; 
and
public -> column-incomplete -> non-existent; 

Turns out we can't. ignore my comment.
<!-- Reviewable comment -K2SrBhNnLmHRz3MHgXD:-1225051090 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
This change has undergone a lot more change. Ready for another review. Thanks!

tamird commented:
Reviewed 1 of 12 files at r1, 7 of 10 files at r2.
Review status: 8 of 9 files reviewed at latest revision, 7 unresolved discussions.

---

<sup>**[sql/descriptor_mutation_test.go, line 21 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2SLnrUC8S0L--Qt3UL)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4f6e4d687f90bc4964f42e8069ffdc63fb0d0013/sql/descriptor_mutation_test.go#L21)):</sup>
"csql" should refer to cockroach/sql, not database/sql.
<!-- Reviewable comment -K2ZxjApgh_2cdIfwoxO:1094975197 -->

---

<sup>**[sql/insert.go, line 54 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2_-4WZ5b38gzlbuqCi)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/insert.go#L54)):</sup>
what's this about?
<!-- Reviewable comment -K2_-4WZ5b38gzlbuqCj:1313663657 -->

---

<sup>**[sql/structured.go, line 58 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2_0EQV6b-V235O50Dt)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/structured.go#L58)):</sup>
why do we need 3 methods? Seems like there are only two that we need: `isVisible` and `isImplicitlyWritable`.

I still think we should have accessors on the TableDescriptor that return slices, though.
<!-- Reviewable comment -K2_0EQV6b-V235O50Du:1470518584 -->

---

<sup>**[sql/structured.proto, line 45 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2Zyb_6YFgswgyM5mNv)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/structured.proto#L45)):</sup>
This is incorrect, there is no need for all the nodes to have transitioned to this state.
<!-- Reviewable comment -K2Zyb_6YFgswgyM5mNw:-1888269894 -->

---

<sup>**[sql/structured.proto, line 50 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2ZyZZG9t4UYmpzmJ78)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/structured.proto#L50)):</sup>
This is similarly incorrect.
<!-- Reviewable comment -K2ZyZZG9t4UYmpzmJ79:1668632298 -->

---

<sup>**[sql/structured.proto, line 51 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2ZydAcKfGGw3wGHW05)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/structured.proto#L51)):</sup>
This is a write+delete state, not delete only.
<!-- Reviewable comment -K2ZydAcKfGGw3wGHW06:454306926 -->

---

<sup>**[sql/update.go, line 142 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2ZzcjUJZht1z8Zgnkk)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/update.go#L142)):</sup>
this seems fragile and difficult to reason about: why are non-readable columns being added to the colIDtoRowIndex map?

Why do you feel that putting this logic into an accessor on the table descriptor is not the way to go?
<!-- Reviewable comment -K2ZzcjUJZht1z8Zgnkl:-433929894 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
I'm working on a further simplification of this change based on what tamir
is saying. It can be further simplified considerably.

On Sat, Nov 7, 2015, 9:58 PM Tamir Duberstein <notifications@github.com>
wrote:

> Reviewed 1 of 12 files at r1, 7 of 10 files at r2.
> Review status: 8 of 9 files reviewed at latest revision, 7 unresolved
> discussions.
> ------------------------------
>
> *sql/descriptor_mutation_test.go, line 21 [r1]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2SLnrUC8S0L--Qt3UL>*
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/4f6e4d687f90bc4964f42e8069ffdc63fb0d0013/sql/descriptor_mutation_test.go#L21>
> ):
>
> "csql" should refer to cockroach/sql, not database/sql.
> ------------------------------
>
> *sql/insert.go, line 54 [r2]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2_-4WZ5b38gzlbuqCi>*
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/insert.go#L54>
> ):
> what's this about?
> ------------------------------
>
> *sql/structured.go, line 58 [r2]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2_0EQV6b-V235O50Dt>*
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/structured.go#L58>
> ):
> why do we need 3 methods? Seems like there are only two that we need:
> isVisible and isImplicitlyWritable.
>
> I still think we should have accessors on the TableDescriptor that return
> slices, though.
> ------------------------------
>
> *sql/structured.proto, line 45 [r2]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2Zyb_6YFgswgyM5mNv>*
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/structured.proto#L45>
> ):
> This is incorrect, there is no need for all the nodes to have transitioned
> to this state.
> ------------------------------
>
> *sql/structured.proto, line 50 [r2]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2ZyZZG9t4UYmpzmJ78>*
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/structured.proto#L50>
> ):
> This is similarly incorrect.
> ------------------------------
>
> *sql/structured.proto, line 51 [r2]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2ZydAcKfGGw3wGHW05>*
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/structured.proto#L51>
> ):
> This is a write+delete state, not delete only.
> ------------------------------
>
> *sql/update.go, line 142 [r2]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2ZzcjUJZht1z8Zgnkk>*
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/update.go#L142>
> ):
> this seems fragile and difficult to reason about: why are non-readable
> columns being added to the colIDtoRowIndex map?
>
> Why do you feel that putting this logic into an accessor on the table
> descriptor is not the way to go?
> ------------------------------
>
> Comments from the review on Reviewable.io
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/3023>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/3023#issuecomment-154772215>
> .
>

vivekmenezes committed:
Proper r/w operations during the addition or deletion of a column

vivekmenezes commented:
Review status: 1 of 7 files reviewed at latest revision, 7 unresolved discussions.

---

<sup>**[sql/insert.go, line 54 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2_-4WZ5b38gzlbuqCi)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/insert.go#L54)):</sup>
Added a comment to explain this. We don't want to add any defaults being added to the length of expected inputs.
<!-- Reviewable comment -K2dX1OqAZQSxzHbU91C:2135844418 -->

---

<sup>**[sql/structured.go, line 58 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2_0EQV6b-V235O50Dt)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/structured.go#L58)):</sup>
removed all these methods.
<!-- Reviewable comment -K2dXMtqRzFeeKqhva3I:1107762333 -->

---

<sup>**[sql/structured.proto, line 45 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2Zyb_6YFgswgyM5mNv)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/structured.proto#L45)):</sup>
I believe this is correct. If you have a node in the WRITE_ONLY state, then while you are garbage collecting the column, the node might receive an insert which will trigger adding a default value which the GC process can miss depending on how far along it's gone.
<!-- Reviewable comment -K2dXQFh0XGrDWml_1pQ:1676052429 -->

---

<sup>**[sql/structured.proto, line 50 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2ZyZZG9t4UYmpzmJ78)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/structured.proto#L50)):</sup>
same comment above applies. If a node is in the DELETE_ONLY state, and a backfill is going on, an insert can come along that doesn't add a default, which the backfill might miss to add depending on how far along it's gone.
<!-- Reviewable comment -K2dXu1azSpd84iduqrX:-1083179146 -->

---

<sup>**[sql/structured.proto, line 51 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2ZydAcKfGGw3wGHW05)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/structured.proto#L51)):</sup>
I think you meant this is write+delete state, not write? Are you suggesting renaming this state? WRITE suggests writing a NULL. 
<!-- Reviewable comment -K2dYVUV1I2Ui7MwzlgN:-1278531995 -->

---

<sup>**[sql/scan.go, line 298 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2SOZmbti5JOBkHU6VZ)** ([raw file](https://github.com/cockroachdb/cockroach/blob/e675a722519a8aa7fbf00d76edfecf5563f58281/sql/scan.go#L298)):</sup>
This has undergone a major restructuring and considerably simplified.
<!-- Reviewable comment -K2dX9X6f97kQ89l0F5w:1992322779 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 1 of 7 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

<sup>**[sql/insert.go, line 75 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2eEvBzxSjxAtqTBQFc)** ([raw file](https://github.com/cockroachdb/cockroach/blob/16a69beaef050638471dea5058b83fab7191c3c0/sql/insert.go#L75)):</sup>
The following code is almost a copy of the above code, but I feel that this code is part of the fast path and probably worth replicating. I don't want to allocate another slice to do this.
<!-- Reviewable comment -K2eEvBzxSjxAtqTBQFd:-1193720682 -->

---

<sup>**[sql/update.go, line 142 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2ZzcjUJZht1z8Zgnkk)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/update.go#L142)):</sup>
Done.
<!-- Reviewable comment -K2eEbJ8QPaLquPfzkJG:-499146009 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Ready for another review. Thanks

tamird commented:
Doesn't DELETE need some special treatment to be consistent with UPDATE?
<!-- Reviewable comment -K2eJv_itJLyNePfBOdz:1021242191 -->

---

Reviewed 7 of 8 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

<sup>**[sql/descriptor_mutation_test.go, line 33 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2eLW_J7rsN2KkbMwxR)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d08a11813c8fd6f649c2aced9af48e1ba9d95c58/sql/descriptor_mutation_test.go#L33)):</sup>
this method's name is quite misleading, since it also mutates the passed-in descriptor.
<!-- Reviewable comment -K2eLW_K5BSPcuDuVA3P:-767337007 -->

---

<sup>**[sql/descriptor_mutation_test.go, line 38 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2eLrEyNoY9QkAxFFVE)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d08a11813c8fd6f649c2aced9af48e1ba9d95c58/sql/descriptor_mutation_test.go#L38)):</sup>
why? shouldn't only the one column be removed?
<!-- Reviewable comment -K2eLrEz-ioxRbN_RWFL:-1102015725 -->

---

<sup>**[sql/descriptor_mutation_test.go, line 86 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2eM0WRkEGfDhIZwAd6)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d08a11813c8fd6f649c2aced9af48e1ba9d95c58/sql/descriptor_mutation_test.go#L86)):</sup>
this is quite misleading, since it doesn't revert at all, but just puts the table into a "migration" state, regardless of whether anything was done at the start of the function.
<!-- Reviewable comment -K2eM0WRkEGfDhIZwAd7:-733645278 -->

---

<sup>**[sql/descriptor_mutation_test.go, line 171 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2eL5QyIziF8V2Q4NzH)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d08a11813c8fd6f649c2aced9af48e1ba9d95c58/sql/descriptor_mutation_test.go#L171)):</sup>
`row...`?
<!-- Reviewable comment -K2eL5QyIziF8V2Q4NzI:1697159726 -->

---

<sup>**[sql/descriptor_mutation_test.go, line 179 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2eLIx1iVJ_xS8yoBro)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d08a11813c8fd6f649c2aced9af48e1ba9d95c58/sql/descriptor_mutation_test.go#L179)):</sup>
use `t.Fatalf` (everywhere)
<!-- Reviewable comment -K2eLIx1iVJ_xS8yoBrp:301798387 -->

---

<sup>**[sql/insert.go, line 75 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2eEvBzxSjxAtqTBQFc)** ([raw file](https://github.com/cockroachdb/cockroach/blob/16a69beaef050638471dea5058b83fab7191c3c0/sql/insert.go#L75)):</sup>
I think a simple closure will let you share most of this code.
<!-- Reviewable comment -K2eJIXmCWtzOdKrsxt-:-2046891754 -->

---

<sup>**[sql/structured.proto, line 45 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2Zyb_6YFgswgyM5mNv)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/structured.proto#L45)):</sup>
Yes, backfill doesn't start until all nodes are in DELETE_ONLY or WRITE_ONLY (depending on direction), but that's not what this comment was saying, and even now it's not totally clear. I'd avoid the term "garbage collection" here.
<!-- Reviewable comment -K2eHHSt7-wKc0NhbJ47:1467453553 -->

---

<sup>**[sql/structured.proto, line 50 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2ZyZZG9t4UYmpzmJ78)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/structured.proto#L50)):</sup>
Ah, I didn't notice "backfilled" before. This one is correct.
<!-- Reviewable comment -K2eHcD70hRqQ8zXc3e-:788470909 -->

---

<sup>**[sql/structured.proto, line 51 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2ZydAcKfGGw3wGHW05)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/structured.proto#L51)):</sup>
Sorry, that's what I meant. Yeah, it might be worth renaming, or at least documenting, that it permits both implicit writes and implicit deletes.
<!-- Reviewable comment -K2eHm9EOksPmNJbfCzl:-305158646 -->

---

<sup>**[sql/structured.proto, line 88 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2eK_r4BlYM5ELoLqpL)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d08a11813c8fd6f649c2aced9af48e1ba9d95c58/sql/structured.proto#L88)):</sup>
There are a bunch of messages involved in this change now. Let's talk tomorrow about how we can make this a bit cleaner.
<!-- Reviewable comment -K2eK_r4BlYM5ELoLqpM:488361464 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
delete uses the rows prefix key to delete all entries.
<!-- Reviewable comment -K2g5a690qt7OFNN9uPd:1766209352 -->

---

Review status: 3 of 7 files reviewed at latest revision, 7 unresolved discussions.

---

<sup>**[sql/descriptor_mutation_test.go, line 33 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2eLW_J7rsN2KkbMwxR)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d08a11813c8fd6f649c2aced9af48e1ba9d95c58/sql/descriptor_mutation_test.go#L33)):</sup>
added a comment
<!-- Reviewable comment -K2g5oXkiiefgS-h836C:-279642405 -->

---

<sup>**[sql/descriptor_mutation_test.go, line 38 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2eLrEyNoY9QkAxFFVE)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d08a11813c8fd6f649c2aced9af48e1ba9d95c58/sql/descriptor_mutation_test.go#L38)):</sup>
There is only one mutation in this descriptor. I've added a check for that condition.
<!-- Reviewable comment -K2g0hykmE0DbQZE0p9e:2104291409 -->

---

<sup>**[sql/descriptor_mutation_test.go, line 86 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2eM0WRkEGfDhIZwAd6)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d08a11813c8fd6f649c2aced9af48e1ba9d95c58/sql/descriptor_mutation_test.go#L86)):</sup>
I've changed this bool to returnWithMutation
<!-- Reviewable comment -K2g0qaFLXWNQnxolw-b:-1433118250 -->

---

<sup>**[sql/descriptor_mutation_test.go, line 171 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2eL5QyIziF8V2Q4NzH)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d08a11813c8fd6f649c2aced9af48e1ba9d95c58/sql/descriptor_mutation_test.go#L171)):</sup>
row is a []string and it expect an []interface type.
<!-- Reviewable comment -K2g5K-F-231PjXlkRNw:181449398 -->

---

<sup>**[sql/structured.proto, line 45 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2Zyb_6YFgswgyM5mNv)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/structured.proto#L45)):</sup>
s/garbage collected/mass deleted/
<!-- Reviewable comment -K2g1o7EFWWOEjq_d1WA:-665068112 -->

---

<sup>**[sql/structured.proto, line 51 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2ZydAcKfGGw3wGHW05)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/structured.proto#L51)):</sup>
added the word implicit to both comments.
<!-- Reviewable comment -K2g2WC9sp6y28YaqvLa:-1395614137 -->

---

<sup>**[sql/structured.proto, line 88 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2eK_r4BlYM5ELoLqpL)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d08a11813c8fd6f649c2aced9af48e1ba9d95c58/sql/structured.proto#L88)):</sup>
ok
<!-- Reviewable comment -K2g5FKy_T2xnjNZ6ie2:5174689 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
@petermattis wanna take a look?

petermattis commented:
Doesn't `Delete` need handling too?
<!-- Reviewable comment -K2hwWPH5N7LkMdjygEd:-1525652796 -->

---

Review status: 2 of 7 files reviewed at latest revision, 10 unresolved discussions.

---

<sup>**[sql/descriptor_mutation_test.go, line 134 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2hvCFzal_2dzUqhF20)** ([raw file](https://github.com/cockroachdb/cockroach/blob/1a6471828ec4ced832e7a89517af48d19d94c331/sql/descriptor_mutation_test.go#L134)):</sup>
Strange for a table to be named `kv` and then to have 3 columns. 
<!-- Reviewable comment -K2hvCFzal_2dzUqhF21:3876944 -->

---

<sup>**[sql/descriptor_mutation_test.go, line 187 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2huufCDo7HBokZEn96)** ([raw file](https://github.com/cockroachdb/cockroach/blob/1a6471828ec4ced832e7a89517af48d19d94c331/sql/descriptor_mutation_test.go#L187)):</sup>
The test cases seem reasonable, but the test itself is hard to follow. I think some more infrastructure might be worthwhile. I'm thinking of something like `lease_test.go` where there are helper functions to acquire, publish and release leases.
<!-- Reviewable comment -K2huufCDo7HBokZEn97:-255587408 -->

---

<sup>**[sql/structured.proto, line 45 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2Zyb_6YFgswgyM5mNv)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/structured.proto#L45)):</sup>
The comment is describing how a schema change operations will use `DELETE_ONLY`, but not the semantics of `DELETE_ONLY`: `The DELETE_ONLY state indicates that the descriptor is visible only to delete operations. Write operations will not add keys for the descriptor and read operations will not see keys for the descriptor.`

Something similar should be added for `WRITE_ONLY`.
<!-- Reviewable comment -K2hs9yW3oBvLn0cGhmE:-132418882 -->

---

<sup>**[sql/table.go, line 34 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2hrlZ9UoLgLUVbWYdt)** ([raw file](https://github.com/cockroachdb/cockroach/blob/1a6471828ec4ced832e7a89517af48d19d94c331/sql/table.go#L34)):</sup>
If this is for testing only, prefix with `Test`. 
<!-- Reviewable comment -K2hrlZ9UoLgLUVbWYdu:1393986885 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 4 files at r5, 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

<sup>**[sql/descriptor_mutation_test.go, line 38 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2eLrEyNoY9QkAxFFVE)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d08a11813c8fd6f649c2aced9af48e1ba9d95c58/sql/descriptor_mutation_test.go#L38)):</sup>
OK. Even with the check, it's better to replace this line with `tableDesc.ColumnMutations = tableDesc.ColumnMutations[1:]`
<!-- Reviewable comment -K2hwv64zNC1Iu4re2Xs:-580213068 -->

---

<sup>**[sql/descriptor_mutation_test.go, line 111 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2hx6Uvxnjez_8FhhmH)** ([raw file](https://github.com/cockroachdb/cockroach/blob/1a6471828ec4ced832e7a89517af48d19d94c331/sql/descriptor_mutation_test.go#L111)):</sup>
how about `tableDesc.ColumnMutations = append(tableDesc.ColumnMutations, csql.ColumnMutation{Column: tableDesc.Columns[len(tableDesc.Columns)-1], Mutation: csql.DescriptorMutation{State: state}}`
<!-- Reviewable comment -K2hx6Uvxnjez_8FhhmI:-212122821 -->

---

<sup>**[sql/structured.proto, line 45 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2Zyb_6YFgswgyM5mNv)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/structured.proto#L45)):</sup>
Minor correction is that those operations will see neither the descriptor's keys nor the descriptor itself.
<!-- Reviewable comment -K2hxYs9cWBKGM9sQY1c:-1088589162 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Delete deletes all the columns and requires no special treatment. In the case of indexes we do require special handling (separate PR). We check that the hidden column is properly deleted in the test by checking that the number of values we see via SQL is equal  to the number of KV values. This gives us the what we have is what we see guarantee.
<!-- Reviewable comment -K2i9n6jha0DmOfmS7X8:970249290 -->

---

Review status: 2 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[sql/descriptor_mutation_test.go, line 134 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2hvCFzal_2dzUqhF20)** ([raw file](https://github.com/cockroachdb/cockroach/blob/1a6471828ec4ced832e7a89517af48d19d94c331/sql/descriptor_mutation_test.go#L134)):</sup>
renamed to t.test
<!-- Reviewable comment -K2i3jQOukU1bc-NAp7w:-764857870 -->

---

<sup>**[sql/descriptor_mutation_test.go, line 187 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2huufCDo7HBokZEn96)** ([raw file](https://github.com/cockroachdb/cockroach/blob/1a6471828ec4ced832e7a89517af48d19d94c331/sql/descriptor_mutation_test.go#L187)):</sup>
Could you elaborate a bit more. 
<!-- Reviewable comment -K2iAg4sZs5_JqUGINXA:-374217705 -->

---

<sup>**[sql/structured.proto, line 45 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2Zyb_6YFgswgyM5mNv)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/structured.proto#L45)):</sup>
I've reworded the comments to define "implicit" and hopefully made it clearer. 
<!-- Reviewable comment -K2i768SOTMq8Ile3zZ2:-663964055 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 1 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[sql/descriptor_mutation_test.go, line 187 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2huufCDo7HBokZEn96)** ([raw file](https://github.com/cockroachdb/cockroach/blob/1a6471828ec4ced832e7a89517af48d19d94c331/sql/descriptor_mutation_test.go#L187)):</sup>
hmmm, perhaps you mean having helper functions to add/delete a mutation that are called outside checkTableData.

BTW, I'm also using some of this infrastructure for the index mutation change and perhaps refactoring within that change is more apt given that I'm already going to refactor for that one.
<!-- Reviewable comment -K2iCZ9zzktXrnkl0rc4:-1889396840 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 1 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[sql/descriptor_mutation_test.go, line 187 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2huufCDo7HBokZEn96)** ([raw file](https://github.com/cockroachdb/cockroach/blob/1a6471828ec4ced832e7a89517af48d19d94c331/sql/descriptor_mutation_test.go#L187)):</sup>
The test is currently table-structured, driven by `testData`. But I think it would be more readable as serial execution of code:

```
t := columnMutationTest{T: testingT}
t.setMutationState("i", DELETE_ONLY)
t.insertRow("a", "z", "q")
t.checkRow(...)
```

You'd have to implement the `columnMutationTest` harness, so the total code for testing might not go down, but I think it would make the test itself more readable. 
<!-- Reviewable comment -K2iDC7seQA97p5cUrtn:2023740176 -->

---

<sup>**[sql/structured.proto, line 45 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2Zyb_6YFgswgyM5mNv)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/structured.proto#L45)):</sup>
Doesn't quite address my concern about the spelling out the semantics of the DELETE_ONLY state.
<!-- Reviewable comment -K2iF5MO8UqlGytGNhBt:-1687730312 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 4 files at r5, 4 of 4 files at r6, 4 of 4 files at r7.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

<sup>**[sql/structured.proto, line 49 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2iFMqxSZo_47SQsQSL)** ([raw file](https://github.com/cockroachdb/cockroach/blob/5ef3091891ab44d840b16390a7494eb765458b07/sql/structured.proto#L49)):</sup>
double space here
<!-- Reviewable comment -K2iFMqxSZo_47SQsQSM:112752660 -->

---

<sup>**[sql/structured.proto, line 94 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2iFYfvS4P7_gP7NBFo)** ([raw file](https://github.com/cockroachdb/cockroach/blob/5ef3091891ab44d840b16390a7494eb765458b07/sql/structured.proto#L94)):</sup>
I wanted to discuss in person, but I missed you today. I think this message can be:
```
message Mutation {
  optional Descriptor descriptor =1 [(gogoproto.nullable) = false];
  optional State state = 2 [(gogoproto.nullable) = false];
}
```
This will let you put indexes in here as well, and remove the need for the extra `DescriptorMutation` message.
<!-- Reviewable comment -K2iFYfwmSwkP1_zUcbU:-464363358 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

<sup>**[sql/structured.proto, line 94 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2iFYfvS4P7_gP7NBFo)** ([raw file](https://github.com/cockroachdb/cockroach/blob/5ef3091891ab44d840b16390a7494eb765458b07/sql/structured.proto#L94)):</sup>
I think you mean, have DescriptorMutation but have a oneof inside it to store an index or column. That's actually quite similar to the original mutation queue proposal in the other PR with a queue of DescriptorMutations. I think the name Mutation is too general for the sql package.
<!-- Reviewable comment -K2iZPX1Mz-KtaLiAArH:2121415494 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

<sup>**[sql/structured.proto, line 94 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2iFYfvS4P7_gP7NBFo)** ([raw file](https://github.com/cockroachdb/cockroach/blob/5ef3091891ab44d840b16390a7494eb765458b07/sql/structured.proto#L94)):</sup>
We already have the `oneof` in `Descriptor`, so I'd just use that instead of duplicating. I'm not married to the name `Mutation`.
<!-- Reviewable comment -K2jFq0-1jhlg8KHjF2s:1291543076 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

<sup>**[sql/structured.proto, line 94 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2iFYfvS4P7_gP7NBFo)** ([raw file](https://github.com/cockroachdb/cockroach/blob/5ef3091891ab44d840b16390a7494eb765458b07/sql/structured.proto#L94)):</sup>
Oh, I'm way wrong. Yes, I'm suggesting a `oneof` as you described.
<!-- Reviewable comment -K2jG7vsTjUX60pTX8T9:1841838599 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 1 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[sql/descriptor_mutation_test.go, line 187 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2huufCDo7HBokZEn96)** ([raw file](https://github.com/cockroachdb/cockroach/blob/1a6471828ec4ced832e7a89517af48d19d94c331/sql/descriptor_mutation_test.go#L187)):</sup>
Added a few functions to make this test more readable. I didn't break up testData, but at least most of the magic is gone.
<!-- Reviewable comment -K2jsf9EeJNVDFxpUNzC:1456887202 -->

---

<sup>**[sql/structured.proto, line 45 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2Zyb_6YFgswgyM5mNv)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7f3f481273994013466ccd314350b853516d455b/sql/structured.proto#L45)):</sup>
Reworded comments above DELETE_ONLY and WRITE_ONLY for both column and index descriptors.
<!-- Reviewable comment -K2i_lIeH90Ukz8c-wRE:1809416894 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Hold on let me make another change to this PR.

On Mon, Nov 9, 2015, 9:41 PM Tamir Duberstein <notifications@github.com>
wrote:

> Review status: all files reviewed at latest revision, 4 unresolved
> discussions.
> ------------------------------
>
> *sql/structured.proto, line 94 [r7]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2iFYfvS4P7_gP7NBFo>*
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/5ef3091891ab44d840b16390a7494eb765458b07/sql/structured.proto#L94>
> ):
>
> Oh, I'm way wrong. Yes, I'm suggesting a oneof as you described.
> ------------------------------
>
> Comments from the review on Reviewable.io
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/3023>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/3023#issuecomment-155265481>
> .
>

vivekmenezes commented:
Review status: 1 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[sql/descriptor_mutation_test.go, line 187 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2huufCDo7HBokZEn96)** ([raw file](https://github.com/cockroachdb/cockroach/blob/1a6471828ec4ced832e7a89517af48d19d94c331/sql/descriptor_mutation_test.go#L187)):</sup>
Got rid of testData. 
<!-- Reviewable comment -K2lWsbCwxN8pQhlP9OS:1952552843 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Ready for another review.

tamird commented:
Reviewed 5 of 6 files at r8.
Review status: 6 of 7 files reviewed at latest revision, 6 unresolved discussions.

---

<sup>**[sql/structured.go, line 138 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2lYCLMOURR83qzBpVH)** ([raw file](https://github.com/cockroachdb/cockroach/blob/cafea2075208ce768f2b876e30b106737df0967d/sql/structured.go#L138)):</sup>
return m.GetDropColumn() and ditto below
<!-- Reviewable comment -K2lYCLMOURR83qzBpVI:420858818 -->

---

<sup>**[sql/structured.go, line 319 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2lYMMz9Y_10fZixJ0k)** ([raw file](https://github.com/cockroachdb/cockroach/blob/cafea2075208ce768f2b876e30b106737df0967d/sql/structured.go#L319)):</sup>
this error should indicate it is even more severe in that there is no descriptor
<!-- Reviewable comment -K2lYMMz9Y_10fZixJ0l:-1691899671 -->

---

<sup>**[sql/structured.proto, line 95 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2jwwZ7efllf-AtPlYk)** ([raw file](https://github.com/cockroachdb/cockroach/blob/fc4d5abc7519a4b99ba806e347f6eca5485163a5/sql/structured.proto#L95)):</sup>
s/queue //
<!-- Reviewable comment -K2jwwZ7efllf-AtPlYl:1531315248 -->

---

<sup>**[sql/structured.proto, line 125 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2jxGVbESy7dVKr71Qo)** ([raw file](https://github.com/cockroachdb/cockroach/blob/fc4d5abc7519a4b99ba806e347f6eca5485163a5/sql/structured.proto#L125)):</sup>
s/ if specified//
<!-- Reviewable comment -K2jxGVbESy7dVKr71Qp:649085228 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 3 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[sql/structured.proto, line 95 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2jwwZ7efllf-AtPlYk)** ([raw file](https://github.com/cockroachdb/cockroach/blob/fc4d5abc7519a4b99ba806e347f6eca5485163a5/sql/structured.proto#L95)):</sup>
reworded to FIFO queue
<!-- Reviewable comment -K2lbIp2JvQc0U_j74J_:-1374336317 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r10.
Review status: 6 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[sql/structured.proto, line 95 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2jwwZ7efllf-AtPlYk)** ([raw file](https://github.com/cockroachdb/cockroach/blob/fc4d5abc7519a4b99ba806e347f6eca5485163a5/sql/structured.proto#L95)):</sup>
No, that's still wrong; the descriptor has just one list of DescriptorMutation, not one for each column or index.
<!-- Reviewable comment -K2m2j8BZaAOhGYv2Iv_:964726288 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 4 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[sql/structured.proto, line 95 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2jwwZ7efllf-AtPlYk)** ([raw file](https://github.com/cockroachdb/cockroach/blob/fc4d5abc7519a4b99ba806e347f6eca5485163a5/sql/structured.proto#L95)):</sup>
Added "containing" . Sorry about that.
<!-- Reviewable comment -K2m6YZvzfyvidOQVPbr:1179616662 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 4 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

<sup>**[sql/descriptor_mutation_test.go, line 125 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2m5-vYJh9X8OSsO-DM)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f013733bae2709e55bfbde063d980e7c8176ca26/sql/descriptor_mutation_test.go#L125)):</sup>
`s/write/writes/g`

`s/DB.s/DB./g`
<!-- Reviewable comment -K2m5-vYJh9X8OSsO-DN:217940373 -->

---

<sup>**[sql/descriptor_mutation_test.go, line 152 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2m5hsJThXZBm5Z-Dcj)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f013733bae2709e55bfbde063d980e7c8176ca26/sql/descriptor_mutation_test.go#L152)):</sup>
I'd prefer something that kept the the schema of the table out of this method. I'd take a page from the logic test here and have something that took an arbitrary query and returned a string composed of the column names and all of the row values, 1 row per line. You can then check the result is what you expected (e.g. 2 columns, 1 row). See `logic_test.go:execQuery`.
<!-- Reviewable comment -K2m5hsJThXZBm5Z-Dck:-544526260 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r11.
Review status: 4 of 7 files reviewed at latest revision, 4 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 3 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[sql/descriptor_mutation_test.go, line 152 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2m5hsJThXZBm5Z-Dcj)** ([raw file](https://github.com/cockroachdb/cockroach/blob/f013733bae2709e55bfbde063d980e7c8176ca26/sql/descriptor_mutation_test.go#L152)):</sup>
I've reorganized this a bit more and added a function that can take a query and expected results and ensure that the query and expected results match up. 
<!-- Reviewable comment -K2n8qG157a9XHXipSlS:300983505 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
ready for another review. Thanks

On Tue, Nov 10, 2015 at 11:11 AM Tamir Duberstein <notifications@github.com>
wrote:

> Reviewed 2 of 2 files at r11.
> Review status: 4 of 7 files reviewed at latest revision, 4 unresolved
> discussions.
> ------------------------------
>
> Comments from the review on Reviewable.io
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/3023>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/3023#issuecomment-155467769>
> .
>

tamird commented:
Reviewed 3 of 4 files at r12.
Review status: 6 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

<sup>**[sql/structured.go, line 293 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2nOjYROU67jJDuIpTu)** ([raw file](https://github.com/cockroachdb/cockroach/blob/1a50ec6114805f5d42468326ec9ea57e17576dea/sql/structured.go#L293)):</sup>
We can use the generated code for better safety here:
```go
unSetEnums := m.State == DescriptorMutation_UNKNOWN || m.Direction == DescriptorMutation_NONE
switch desc := m.GetDescriptor_().(type) {
  case *DescriptorMutation_Column:
    if unSetEnums {
      col := desc.Column
      return util.Errorf("mutation in state %s, direction %s, col %s, id %v", m.State, m.Direction, col.Name, col.ID)
    }
  case *DescriptorMutation_Index:
    if unSetEnums {
      idx := desc.Index
      return util.Errorf("mutation in state %s, direction %s, col %s, id %v", m.State, m.Direction, idx.Name, idx.ID)
    }
  default:
    return util.Errorf("mutation in state %s, direction %s, and no column/index descriptor", m.State, m.Direction)
}
```
<!-- Reviewable comment -K2nOjYROU67jJDuIpTv:1991366952 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

tamird commented:
I find the test here quite hard to read, still. LGTM otherwise though!
<!-- Reviewable comment -K2s6XZmntzq5tJxLCZy:-864533816 -->

---

Reviewed 2 of 3 files at r13.
Review status: 6 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

<sup>**[sql/descriptor_mutation_test.go, line 164 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2s5hs2uavtC9ySsj91)** ([raw file](https://github.com/cockroachdb/cockroach/blob/c500cf45cb878690e1405f448d58fe8daae72e29/sql/descriptor_mutation_test.go#L164)):</sup>
nit: we have a nice pattern that we use elsewhere, which doesn't export the actual boolean but only a function that sets it to true and returns a function that sets it back to false. Then we invoke it as `defer TestDisableTableLeases()()`
<!-- Reviewable comment -K2s5hs2uavtC9ySsj92:418327536 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
That's good to know. I've sprinkled a lot of comments to make it easier to understand, and renamed dropMutations to makeMutationsLive which is easier to read. I'll be merging this PR soon. If you have any comments on the new changes please add them in here and I'll add them to the other PR on index mutations.
<!-- Reviewable comment -K2tZWBQAEF9DHw4RkPb:-233884308 -->

---

Review status: 3 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

<sup>**[sql/descriptor_mutation_test.go, line 164 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023#-K2s5hs2uavtC9ySsj91)** ([raw file](https://github.com/cockroachdb/cockroach/blob/c500cf45cb878690e1405f448d58fe8daae72e29/sql/descriptor_mutation_test.go#L164)):</sup>
nice
<!-- Reviewable comment -K2tNqHSkWkXTzYNcYLa:677880290 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3023)
<!-- Sent from Reviewable.io -->
